packages/celotool/src/e2e-tests/utils.ts

```diff
@@ -385,109 +125,155 @@ export async function migrateContracts(
   const mnemonic =
     'jazz ripple brown cloth door bridge pen danger deer thumb cable prepare negative library vast'
   const validatorInstances = gethConfig.instances.filter((x: any) => x.validating)
+
   const numValidators = validatorInstances.length
+
   const validatorPrivateKeys = getPrivateKeysFor(AccountType.VALIDATOR, mnemonic, numValidators)
   const attestationKeys = getPrivateKeysFor(AccountType.ATTESTATION, mnemonic, numValidators)
-  const validators = getValidators(mnemonic, numValidators)
+  const validators = getValidatorsInformation(mnemonic, numValidators)
+
+  const proxyInstances = gethConfig.instances.filter((x: any) => x.isProxy)
+  const numProxies = proxyInstances.length
+
+  const proxyPrivateKeys = getPrivateKeysFor(AccountType.PROXY, mnemonic, numProxies)
+  const proxyEnodes = proxyPrivateKeys.map((x: any, i: number) => [
+    proxyInstances[i].name,
+    getEnodeAddress(privateKeyToPublicKey(x), '127.0.0.1', proxyInstances[i].proxyport!),
+    getEnodeAddress(privateKeyToPublicKey(x), '127.0.0.1', proxyInstances[i].port),
+  ])
+
   const argv = require('minimist')(process.argv.slice(2))
   const branch = argv.branch || 'master'
-  const gethRepoPath = argv.localgeth || '/tmp/geth'
-  const gethBinaryPath = `${gethRepoPath}/build/bin/geth`
+
+  gethConfig.gethRepoPath = argv.localgeth || '/tmp/geth'
+  const gethBinaryPath = `${gethConfig.gethRepoPath}/build/bin/geth`
+  const bootnodeBinaryPath = `${gethConfig.gethRepoPath}/build/bin/bootnode`
 
   const before = async () => {
-    if (!argv.localgeth) {
-      await checkoutGethRepo(branch, gethRepoPath)
+    if (!argv.localgeth && gethConfig.gethRepoPath) {
+      await checkoutGethRepo(branch, gethConfig.gethRepoPath)
+    }
+
+    if (gethConfig.gethRepoPath) {
+      await buildGeth(gethConfig.gethRepoPath)
+    }
+
+    if (!gethConfig.keepData && fs.existsSync(gethConfig.runPath)) {
+      await resetDataDir(gethConfig.runPath, verbose)
+    }
+
+    if (!fs.existsSync(gethConfig.runPath)) {
+      // @ts-ignore
+      fs.mkdirSync(gethConfig.runPath, { recursive: true })
     }
-    await buildGeth(gethRepoPath)
-    await setupTestDir(TEST_DIR)
-    await writeGenesis(validators, GENESIS_PATH, gethConfig.genesisConfig)
+
+    await writeGenesis(gethConfig, validators, verbose)
+
+    let bootnodeEnode: string = ''
+
+    if (gethConfig.useBootnode) {
+      bootnodeEnode = await startBootnode(bootnodeBinaryPath, mnemonic, gethConfig, verbose)
+    }
+
     let validatorIndex = 0
+    let proxyIndex = 0
+
     for (const instance of gethConfig.instances) {
+      // Non proxied validators and proxies should connect to the bootnode
+      if ((instance.validating && !instance.isProxied) || instance.isProxy) {
+        if (gethConfig.useBootnode) {
+          instance.bootnodeEnode = bootnodeEnode
+        }
+      } else if (instance.validating && instance.isProxied) {
```

Comment:
If you apply the above suggestion, then this is just `else`
```suggestion
      } else {
```